src/core/internal/atomic.d

```diff
@@ -12,664 +12,1075 @@ module core.internal.atomic;
+                return *cast(typeof(return)*)&res;
+            }
+            else static if (T.sizeof == short.sizeof)
+            {
+                ushort res = __atomic_exchange_2(cast(shared)dest, *cast(ushort*)&value, order);
+                return *cast(typeof(return)*)&res;
+            }
+            else static if (T.sizeof == int.sizeof)
+            {
+                uint res = __atomic_exchange_4(cast(shared)dest, *cast(uint*)&value, order);
+                return *cast(typeof(return)*)&res;
+            }
+            else static if (T.sizeof == long.sizeof && GNU_Have_64Bit_Atomics)
+            {
+                ulong res = __atomic_exchange_8(cast(shared)dest, *cast(ulong*)&value, order);
+                return *cast(typeof(return)*)&res;
+            }
+            else static if (GNU_Have_LibAtomic)
+            {
+                T res = void;
+                __atomic_exchange(T.sizeof, cast(shared)dest, cast(void*)&value, &res, order);
+                return res;
+            }
+            else
+                static assert(0, "Invalid template type specified.");
+        }
+        else
+        {
+            getAtomicMutex.lock();
+            scope(exit) getAtomicMutex.unlock();
 
-            Lcpuid:
+            T res = *dest;
+            *dest = value;
+            return res;
+        }
+    }
 
-                // Slow path: We use cpuid to serialize. This is
-                // significantly slower than mfence, but is the
-                // only serialization facility we have available
-                // on older non-SSE2 chips.
-                push EBX;
+    bool atomicCompareExchangeWeak(MemoryOrder succ = MemoryOrder.seq, MemoryOrder fail = MemoryOrder.seq, T)(T* dest, T* compare, T value) pure nothrow @nogc @trusted
+        if (CanCAS!T)
+    {
+        return atomicCompareExchangeImpl!(succ, fail, true)(dest, compare, value);
+    }
 
-                mov EAX, 0;
-                cpuid;
+    bool atomicCompareExchangeStrong(MemoryOrder succ = MemoryOrder.seq, MemoryOrder fail = MemoryOrder.seq, T)(T* dest, T* compare, T value) pure nothrow @nogc @trusted
+        if (CanCAS!T)
+    {
+        return atomicCompareExchangeImpl!(succ, fail, false)(dest, compare, value);
+    }
 
-                pop EBX;
+    bool atomicCompareExchangeStrongNoResult(MemoryOrder succ = MemoryOrder.seq, MemoryOrder fail = MemoryOrder.seq, T)(T* dest, const T compare, T value) pure nothrow @nogc @trusted
+        if (CanCAS!T)
+    {
+        return atomicCompareExchangeImpl!(succ, fail, false)(dest, cast(T*)&compare, value);
+    }
 
-            Lend:
+    private bool atomicCompareExchangeImpl(MemoryOrder succ = MemoryOrder.seq, MemoryOrder fail = MemoryOrder.seq, bool weak, T)(T* dest, T* compare, T value) pure nothrow @nogc @trusted
+        if (CanCAS!T)
+    {
+        bool res = void;
 
-                ret;
+        static if (GNU_Have_Atomics || GNU_Have_LibAtomic)
```

Comment:
3 out if 5 lines in this section is wasted vertical space :/
Most of the lines here are `{`, `}` and the line split, which is split well before the horiz white space even github shows here in the review render ;)

The 3 if/else blocks up a little waste 50% if lines to `{`, `}`, but this blocks triggers me ;)
Where I'm from, monitors have been 16/9 for almost at least 15 years... so I feel like vertical space is about 16/9th's more valuable ;)

Reply:
https://dlang.org/dstyle

Reply:
I probably have a subconscious habit of keeping to 80 chars as well.

Reply:
Any though it's just an opinion, I don't personally find compressing the format very readable.

![image](https://user-images.githubusercontent.com/397929/66773212-d9654380-eebe-11e9-81ca-5d40d97a7060.png)

Reply:
Removed some braces just to make @TurkeyMan happy.

Reply:
left the awkward line breaks though :P

Reply:
It's there for readability. :-P